File: baekjoon/1707.py
# input
# 2
# 3 2
# 1 3
# 2 3
# 4 4
# 1 2
# 2 3
# 3 4
# 4 2
# output
# YES
# NO
from collections import defaultdict
import sys
input = sys.stdin.readline
K = int(input())

def solution(edges, nodes, start):
    color = 1
    nodes[start] = color
    stack = [(i, nodes[i])]
    while stack:
        curr_node, curr_color = stack.pop()
        adjacents = edges[curr_node]
        for a in adjacents:
            if nodes[a] == curr_color:
                return False
            elif nodes[a] == 0:
                nodes[a] = curr_color*(-1)
                stack.append((a, nodes[a]))
            else:
                continue
    return True

for _ in range(K):
    V, E = map(int, input().split(" "))
    edges = [[] for _ in range(V)]
    for _ in range(E):
        n1, n2 = map(int, input().split(" "))
        edges[n1-1].append(n2-1)
        edges[n2-1].append(n1-1)
    visited = [0 for _ in range(V)]
    for i in range(V):
        if visited[i] == 0:
            if not solution(edges, visited, i):
                print("NO")
                break
    else:
        print("YES")

File: baekjoon/gold/1707/1707.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	reader *bufio.Reader = bufio.NewReader(os.Stdin)
	writer *bufio.Writer = bufio.NewWriter(os.Stdout)
)

func main() {
	var k, v, e, n1, n2 int

	defer writer.Flush()
	fmt.Fscanf(reader, "%d\n", &k)

	for i := 0; i < k; i++ {
		fmt.Fscanf(reader, "%d %d\n", &v, &e)
		edges := make([][]int, v)
		for j := 0; j < e; j++ {
			fmt.Fscanf(reader, "%d %d\n", &n1, &n2)
			edges[n1-1] = append(edges[n1-1], n2-1)
			edges[n2-1] = append(edges[n2-1], n1-1)
		}
		visited := make([]int, v)
		for i := 0; i < v; i++ {
			if visited[i] == 0 && !solution(edges, visited, i) {
				goto NO
			}
		}
		fmt.Fprintln(writer, "YES")
		continue
	NO:
		fmt.Fprintln(writer, "NO")
	}

}

func solution(e [][]int, v []int, s int) bool {
	var stack [][]int
	color := 1
	v[s] = color
	stack = append(stack, []int{s, v[s]})
	for len(stack) > 0 {
		p := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		currNode, currColor := p[0], p[1]
		adj := e[currNode]
		for _, ae := range adj {
			if v[ae] == currColor {
				return false
			} else if v[ae] == 0 {
				v[ae] = -currColor
				stack = append(stack, []int{ae, v[ae]})
			}
		}
	}
	return true
}
